### ISSUE TITLE
Likelihood/Posterior mismatch in multi-threaded mode

### ISSUE THREAD

**Issue by oschulz** [bat/bat#205]
When running multi-threaded, with a fast likelihood function, there are rare mismatches between likelihood and posterior (at least in the chain output):

![c1](https://cloud.githubusercontent.com/assets/546147/24396588/ff896bc0-13a3-11e7-90c1-47ed7bc6e3cb.png)
![c1-2](https://cloud.githubusercontent.com/assets/546147/24396589/ff9e5d28-13a3-11e7-9d04-cf4bb34a8537.png)
![c1-3](https://cloud.githubusercontent.com/assets/546147/24396590/ffa841ee-13a3-11e7-8b42-b50f5b18c724.png)

Conditions: Flat prior, 14 chains on 14 threads, each with 1000000 iterations.

**Comment by fredRos** [comment]
I'm confident the log_posterior is computed correctly. So I presume the error might be in a race condition in https://github.com/bat/bat/blob/master/src/BCEngineMCMC.cxx#L599 `GetCurrentChain` that is used in https://github.com/bat/bat/blob/master/src/BCModel.cxx#L86 such that the wrong likelihood value is stored.

I should reproduce on a simple test problem with and without treads and fix the seeds just like in the unit test to make sure `logprop` is always correct and that the error is in `GetCurrentChain`.

**Comment by fredRos** [comment]
I'm not sure why `UpdateChainIndex` is also called in `Remarginalize` but that seems to be independent.

**Comment by greenwald** [comment]
Coincidentally, @tpoeschl and I discovered a bug last week that may be this very same one. We solved it---but I'm not sure if Thomas has finished testing that it is truly fixed. The chain index is cleared too early---or, alternatively, `UpdateChainIndex` is not called enough. We parallelize over the proposing of new points and then after all chains have had their new points selected, we call `EvaluateObservables`, looping over the chains again in the process. We need to call `UpdateChainIndex` in `EvaluateObservables` for the user to properly be able to query the current chain in their `CalculateObservables`.

`Remarginalize` calls this function because it may also have to call `CalculateObservables`. (I suppose it's enough then to only call it in `EvaluateObservables`.)

We also only put the current points into the tree (serially) after we have selected them (in parallel). Since fMCMCChain and the current chain index don't have to match in parallel, it is probably this that is causing the mismatch. We would have to call `UpdateChainIndex` also in the loop of filling the tree, and then use `GetCurrentChain` to access the correct entries in all relevant vectors for filling the tree.

**Comment by fredRos** [comment]
I see, it seems these two issues are related. What you say supports my
suspicion that only the storing of likelihood values is incorrect but
that the chains perform correctly

I'm looking forward to a pull request!

**Comment by fredRos** [comment]
@greenwald @tpoeschl Can we get together to merge your fixes into master? I want to fix outstanding bugs for 1.0 until mid April. It seems you have already done the work, so it would be great if you could file a pull request. I can help with resolving conflicts
